File: src/com/kingmang/lino/symbol/Environment.java
package com.kingmang.lino.symbol;

import java.util.HashMap;
import java.util.Hashtable;

public class Environment {
    private final Environment previous;
    private final HashMap<String, Symbol> table;

    public Environment(Environment previous){
        table = new HashMap<>();
        this.previous = previous;
    }

    public void put(String str, Symbol symbol){
        table.put(str, symbol);
    }

    public Symbol get(String str){
        for(Environment environment = this; environment != null; environment = environment.previous){
            Symbol found = (environment.table.get(str));
            if(found != null) return found;
        }
        return null;
    }
}

File: symbol/environment.go
package symbol

type Environment struct {
	table    map[string]*Symbol
	previous *Environment
}

func NewEnvironment(previous *Environment) *Environment {
	return &Environment{
		table:    make(map[string]*Symbol),
		previous: previous,
	}
}

func (e *Environment) Put(name string, sym *Symbol) {
	e.table[name] = sym
}

func (e *Environment) Get(name string) *Symbol {
	for env := e; env != nil; env = env.previous {
		if found, ok := env.table[name]; ok {
			return found
		}
	}
	return nil
}
